### PR TITLE
Network implement

### PR DESCRIPTION
**Pull request by dzhwinter**

### CODE REVIEW

File: paddle/framework/net.h [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+ *
+ * This is the base class of network, all the networks should implement the apis
+ * it defines.
+ */
+class Net {
+ public:
+  /**
+   * @brief Infer shapes of all inputs and outputs of operators.
+   */
+  virtual bool InferShape(Scope *scope) override;
```
Comment (Superjom) [inline paddle/framework/net.h]: 这些  bool 都改成 Error 吧

File: paddle/framework/net.cc [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace paddle {
+namespace framework {
+
+PlainNet::PlainNet(const NetDesc& def) {}
+
+Error PlainNet::InferShape(Scope* scope) {
+  for (auto& op : ops_) {
+    // wrong shape
+    auto err = op.InferShape();
```
Comment (reyoung) [inline paddle/framework/net.cc]: We should not use `Error` now, please use `PADDLE_ENFORCE` instead.

**Comment by Superjom** [comment]
Several interfaces may be needed:

RNN need a proto or string as proto to create a `StepNet`, so need
```c++
void Serialize(NetDesc* net_desc) = 0
```

Forward or backward using `Net.Run` seems difficult because `end_idx` is hard to determine, so following methods may be needed:

```c++
void Forward() = 0;
void Backward() = 0;
```

**Comment by Superjom** [comment]
Do `AddOptimizerOps` and `AddBackwardOps` must be called in order? 
`AddBackwardOps` first, `AddOptimizerOps` second?

@dzhwinter @reyoung

**Comment by jacquesqiao** [comment]
AddOptimizerOps should call after AddBackwardOps

**Comment by QiJune** [comment]
The AddBackwardOps will add all backward related operators. So, we have to find a way to tell the input and output variables of each backward operator.
And the Optimizer will take parameter/gradient/momentum/... as input and update the parameter.
The Optimizer operator will take calculate one pair of parameter and gradient. But the optimizer will take all.
Since, we also have to make checkpoint of a Net parameters, I suggest we put forward a concept called ```Model```.
The ```Model``` has some maps which records the relationship of parameters/gradients. And learning rate, batch_step can also be found in ```Model```. 
The ```Model``` has a  method called ```CheckPoint```, and will serialize its data member to disk.

**Comment by Superjom** [comment]
After a discussion, the current decision is as follows:

- add a temporary solution for Net's backward @dzhwinter 
- no need to add `Serialize` currently, use `AddOp` to build a step net for RNN.
- `Model` is a user-related API, currently is not needed.

**Comment by dzhwinter** [comment]
check in for @Superjom 's rnn op dependency.

**Pull request merged**